
# Data Quality Assessment: Sprocket Central Pty Ltd

### Table of Contents
<ul>
    <li><a href="#intro">Introduction</a></li>
    <li><a href="#setting">Setting Data</a></li>
    <li><a href="#assessing">Assessing Data</a></li>
    <li><a href="#issues">Identified Issues</a></li>
    <li><a href="#recommendations">Recommendations</a></li>
</ul>

<a id='intro'></a>
## Introduction

Sprocket Central Pty Ltd , a medium size bikes & cycling accessories organisation, has approached Tony Smith (Partner) in KPMG’s Lighthouse & Innovation Team. Sprocket Central Pty Ltd  is keen to learn more about KPMG’s expertise in its Analytics, Information & Modelling team. They need help with its customer and transactions data. The organisation has a large dataset relating to its customers, but their team is unsure how to effectively analyse it to help optimise its marketing strategy.

The purpose of this assessment is to optimised the customer's datasets, so it can have better quality for analysis, and various insights can be uncovered.


The process of gathering the datasets required for assessment wass through the file provided by the company, [here](https://cdn-assets.theforage.com/vinternship_modules/kpmg_data_analytics/KPMG_VI_New_raw_data_update_final.xlsx).
A quick visual assessment using Google Sheet revealed that the three datasets we would be working with were present in different tab, so I proceeded to download these three datasets as a CSV, in order to begin my assessment.

I would be assessing the 3 datasets extracted from the provided sheet.
1. Customer Demographic
2. Customer Address
3. Transactions Data in the past 3 months

<a id='setting'></a>
## Setting Data

The goal here is to import all the 3 datasets and load into pandas dataframes, so I can perform further visual assessment and also programmatic assessment.

In [1]:
# Importing the pandas package.

import pandas as pd

In [2]:
'''
    Loading our datasets into pandas dataframes.
    The method of loading all 3 datasets is the same.
'''

customer_demographic = pd.read_csv('KPMG_Task_CustomerDemographic.csv', header=1);
customer_addresses = pd.read_csv('KPMG_Task_CustomerAddresses.csv', header=1);
transactions = pd.read_csv('KPMG_Task_Transactions.csv', header=1);

<a id='assessing'></a>
## Assessing Data

Now that are having our 3 datasets represented in 3 different pandas dataframes, we want to perform both Visual and Programmatic assessment of our gathered datasets. This will enable us identify any possible data qaulity or data tidiness issues and document it.


Our datasets

1. `customer_demographic`
2. `customer_addresses`
3. `transactions`

#### 1. `customer_demogrpahic`

In [3]:
# A quick overview of the Customer Demographic dataset

customer_demographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases         DOB               job_title  \
0                                   93  1953-10-12     Executive Secretary   
1                                   81  1980-12-16  Administrative Officer   
2                                   61  1954-01-20      Recruiting Manager   
3                                   33  1961-10-03                     NaN   
4                                   56  1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                              1-Feb      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

`customer_demographic` columns:
- **customer_id**: The unique identifier for each customer
- **first_name**: The assgined first name of each customer
- **last_name**: The assigned last name of each customer
- **gender**: The assigned gender of each customer
- **past_3_years_bike_related_purchases**: Number of bike related purchases the customer has made in the last 3 years
- **DOB**: The date of birth of each customer (year/month/day)
- **job_title**: The job title for each customer
- **job_industry_category**: The corresponding industry in which the customer works
- **wealth_segment**: Categorised assignment based on the customer's wealth
- **deceased_indicator**: An indication for whether the customer is deceased
- **default**: -
- **owns_car**: An indication for whether the custoemr has a car
- **tenure**: -

In [4]:
# An wholistic view of the columns and their datatypes

customer_demographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   customer_id                          4000 non-null   int64  
 1   first_name                           4000 non-null   object 
 2   last_name                            3875 non-null   object 
 3   gender                               4000 non-null   object 
 4   past_3_years_bike_related_purchases  4000 non-null   int64  
 5   DOB                                  3913 non-null   object 
 6   job_title                            3494 non-null   object 
 7   job_industry_category                3344 non-null   object 
 8   wealth_segment                       4000 non-null   object 
 9   deceased_indicator                   4000 non-null   object 
 10  default                              3698 non-null   object 
 11  owns_car                      

In [5]:
customer_demographic.describe()

customer_id  past_3_years_bike_related_purchases       tenure
count  4000.000000                          4000.000000  3913.000000
mean   2000.500000                            48.890000    10.657041
std    1154.844867                            28.715005     5.660146
min       1.000000                             0.000000     1.000000
25%    1000.750000                            24.000000     6.000000
50%    2000.500000                            48.000000    11.000000
75%    3000.250000                            73.000000    15.000000
max    4000.000000                            99.000000    22.000000

In [6]:
'''
    Due to the observation of inconsistent values in the gender columns,
    I proceeded to check for all availabe value types in the column.
'''

customer_demographic['gender'].value_counts()

Female    2037
Male      1872
U           88
F            1
Femal        1
M            1
Name: gender, dtype: int64

In [7]:
# Checking for the oldest customer in the dataset

customer_demographic.sort_values(by="DOB")

customer_id first_name  last_name  gender  \
33             34   Jephthah   Bachmann       U   
719           720     Darrel      Canet    Male   
1091         1092     Katlin    Creddon  Female   
3409         3410    Merrili    Brittin  Female   
2412         2413      Abbey     Murrow    Male   
...           ...        ...        ...     ...   
3778         3779      Ulick     Daspar       U   
3882         3883      Nissa     Conrad       U   
3930         3931      Kylie      Epine       U   
3934         3935     Teodor  Alfonsini       U   
3997         3998     Sarene    Woolley       U   

      past_3_years_bike_related_purchases         DOB  \
33                                     59  1843-12-21   
719                                    67  1931-10-23   
1091                                   56  1935-08-22   
3409                                   93  1940-09-22   
2412                                   27  1943-08-11   
...                                   ...         ...   
3778                                   68         NaN   
3882                                   35         NaN   
3930                                   19         NaN   
3934                                   72         NaN   
3997                                   60         NaN   

                     job_title job_industry_category     wealth_segment  \
33             Legal Assistant                    IT  Affluent Customer   
719         Recruiting Manager                Retail  Affluent Customer   
1091        VP Quality Control                Retail      Mass Customer   
3409                       NaN              Property      Mass Customer   
2412  Environmental Specialist         Manufacturing     High Net Worth   
...                        ...                   ...                ...   
3778                       NaN                    IT  Affluent Customer   
3882           Legal Assistant                    IT      Mass Customer   
3930                       NaN                    IT     High Net Worth   
3934                       NaN                    IT     High Net Worth   
3997         Assistant Manager                    IT     High Net Worth   

     deceased_indicator             default owns_car  tenure  
33                    N                 NaN       No    20.0  
719                   N  1;DROP TABLE users       No     6.0  
1091                  N  ì¬íê³¼íì ì´íì°êµ¬ì       No     5.0  
3409                  N                   á       No    16.0  
2412                  N                  á       Yes    17.0  
...                 ...                 ...      ...     ...  
3778                  N                 NaN       No     NaN  
3882                  N                 NaN       No     NaN  
3930                  N                 NaN      Yes     NaN  
3934                  N                 NaN      Yes     NaN  
3997                  N                 NaN       No     NaN  

[4000 rows x 13 columns]

In [8]:
customer_demographic['deceased_indicator'].value_counts()

N    3998
Y       2
Name: deceased_indicator, dtype: int64

#### 2. `customer_addresses`

In [9]:
# A quick overview of the Customer Addresses dataset

customer_addresses.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

`customer_addresses` columns:
- **customer_id**: The unique identifier for each customer
- **address**: The main address for each customer
- **postcode**: The corresponding postal code of the customer's address
- **state**: The correspoinding state for the main address of each customer
- **country**: The correspoinding country for the main address of each customer
- **property_valuation**: -

In [10]:
# An wholistic view of the columns and their datatypes

customer_addresses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [11]:
customer_addresses.describe()

customer_id     postcode  property_valuation
count  3999.000000  3999.000000         3999.000000
mean   2003.987997  2985.755939            7.514379
std    1154.576912   844.878364            2.824663
min       1.000000  2000.000000            1.000000
25%    1004.500000  2200.000000            6.000000
50%    2004.000000  2768.000000            8.000000
75%    3003.500000  3750.000000           10.000000
max    4003.000000  4883.000000           12.000000

In [12]:
# Customers without address information

customer_demographic[~customer_demographic['customer_id'].isin(customer_addresses['customer_id'])]

customer_id first_name last_name  gender  \
2             3      Arlin    Dearle    Male   
9            10   Fiorenze   Birdall  Female   
21           22    Deeanne  Durtnell  Female   
22           23       Olav     Polak    Male   

    past_3_years_bike_related_purchases         DOB                job_title  \
2                                    61  1954-01-20       Recruiting Manager   
9                                    49  1988-10-11  Senior Quality Engineer   
21                                   79  1962-12-10                      NaN   
22                                   43  1995-02-10                      NaN   

   job_industry_category  wealth_segment deceased_indicator default owns_car  \
2               Property   Mass Customer                  N   1-Feb      Yes   
9     Financial Services   Mass Customer                  N    ð©ð½      Yes   
21                    IT   Mass Customer                  N  ï¼ï¼ï¼       No   
22                   NaN  High Net Worth                  N   1-Feb      Yes   

    tenure  
2     15.0  
9     20.0  
21    11.0  
22     1.0

#### 3. `transactions`

In [13]:
# A quick overview of the Transactions dataset

transactions.head()

transaction_id  product_id  customer_id transaction_date online_order  \
0               1           2         2950        25/2/2017        False   
1               2           3         3120        21/5/2017         True   
2               3          37          402       16/10/2017        False   
3               4          88         3135        31/8/2017        False   
4               5          78          787        1/10/2017         True   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price standard_cost  product_first_sold_date  
0       71.49        $53.62                  41245.0  
1     2091.47       $388.92                  41701.0  
2     1793.43       $248.82                  36361.0  
3     1198.46       $381.10                  36145.0  
4     1765.30       $709.48                  42226.0

`transactions` columns:

The title of the columns in this dataset makes it pretty straightforward to deduce when working with it.

In [14]:
# An wholistic view of the columns and their datatypes

transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   transaction_id           20000 non-null  int64  
 1   product_id               20000 non-null  int64  
 2   customer_id              20000 non-null  int64  
 3   transaction_date         20000 non-null  object 
 4   online_order             19640 non-null  object 
 5   order_status             20000 non-null  object 
 6   brand                    19803 non-null  object 
 7   product_line             19803 non-null  object 
 8   product_class            19803 non-null  object 
 9   product_size             19803 non-null  object 
 10  list_price               20000 non-null  float64
 11  standard_cost            19803 non-null  object 
 12  product_first_sold_date  19803 non-null  float64
dtypes: float64(2), int64(3), object(8)
memory usage: 2.0+ MB


In [15]:
transactions.describe()

transaction_id   product_id   customer_id    list_price  \
count    20000.000000  20000.00000  20000.000000  20000.000000   
mean     10000.500000     45.36465   1738.246050   1107.829449   
std       5773.647028     30.75359   1011.951046    582.825242   
min          1.000000      0.00000      1.000000     12.010000   
25%       5000.750000     18.00000    857.750000    575.270000   
50%      10000.500000     44.00000   1736.000000   1163.890000   
75%      15000.250000     72.00000   2613.000000   1635.300000   
max      20000.000000    100.00000   5034.000000   2091.470000   

       product_first_sold_date  
count             19803.000000  
mean              38199.776549  
std                2875.201110  
min               33259.000000  
25%               35667.000000  
50%               38216.000000  
75%               40672.000000  
max               42710.000000

In [16]:
transactions['online_order'].value_counts()

True     9829
False    9811
Name: online_order, dtype: int64

In [17]:
transactions['order_status'].value_counts()

Approved     19821
Cancelled      179
Name: order_status, dtype: int64

In [18]:
transactions['brand'].value_counts()

Solex             4253
Giant Bicycles    3312
WeareA2B          3295
OHM Cycles        3043
Trek Bicycles     2990
Norco Bicycles    2910
Name: brand, dtype: int64

In [19]:
transactions['product_line'].value_counts()

Standard    14176
Road         3970
Touring      1234
Mountain      423
Name: product_line, dtype: int64

In [20]:
transactions['product_class'].value_counts()

medium    13826
high       3013
low        2964
Name: product_class, dtype: int64

In [21]:
transactions['product_size'].value_counts()

medium    12990
large      3976
small      2837
Name: product_size, dtype: int64

<a id='issues'></a>
## Identified Issues

During the process of assessing the various datasets, we identified various issues with the quality of the datasets. And for the purpose of documentations, we use a Quality and Tidiness Issues Framework. Below are the documented Quality and Tidiness issues as it pertains to each datasets. 

> Quality Issues - This is when a dataset has issues with its content, like missing data, invalid data, inaccurate data, inconsistent data.
> Tidiness Issues - This is when a dataset has issues with its structure, like unconventional column names, datatypes, etc 

### Quality Issues

`customer_demographic`
- last_name: missing values
- gender: Full gender sometimes (Male, Female), abbreviations other times (M, F), 
- DOB: missing values
- DOB has a string datatype
- job_title: missing values
- job_industry_category: missing values
- default: unrecognised data format
- tenure: missing values

`customer_addresses`
- missing record (3999 instead of 4000)


`transactions`
- transaction_date is a string
- online_order: missing values
- brand: missign values
- product_line: missing values
- product_class: missing values
- product_size: missing values
- standard_cost: missing values
- standard_cost: has a string datatype
- product_first_sold_date: missing values


### Tidiness Issues

`customer_demographic`
- DOB - does not follow the lowercase naming format of other columns
- Unrealistic DOB for a customer with (1843-12-21)

`transactions`
- transaction_date format should consistent (year-month-day)
- online_order has a datatype string [with True and False values]
- [order_status, brand, product_line, product_class, product_size] all have datatype string

One table `customers` split into two `customer_demographic` and `customer_addresses`


<a id='recommendations'></a>
## Recommendations

After thorough assessment of the provided datasets, I identified both quality and tidiness issues which have been properly documented. The next step is to give recommendations on how the datasets can be cleaned, so it is ready for both exploratory analysis and explanatory analysis.

#### Quality Issues:

`customer_demographic`
- The empty values in the last_name can be replace with N/A (Not Applicable).
- For consistency, the gender column needs to be reformatted to use a specific representation of gender (Male, Female, Others) or (M, F, O)
- The datatype of the DOB column should be converted to datetime
- When it is time for analysis, if the DOB column is essential, then entries without DOB should be dropped.
- The entries with empty job_title and job_industry_category should be droppoed if it would be crucial to our analysis.
- The default column should be removed, it is not clear enough as to what it represents or how it can be used.
- The entries with empty tenure should be droppoed if it would be crucial to our analysis.

`customer_addresses`
- There are 4 customers without address information. We can drop these entries.


`transactions`
- The transaction_date column datatype should be converted to a datetime
- The entries with missing online_order values should be dropped and datatype converted to boolean
- The entries with missing brand should be dropped and datatype should be converted to categorical
- The entries with missing product_line should be dropped and datatype should be converted to categorical
- The entries with missing product_class should be dropped and datatype should be converted to categorical
- The entries with missing product_size should be dropped and datatype should be converted to categorical
- The entries with missing standard_cost should be dropped, data in column should be reformatted to just the numeric value (striping the currency sign)
- The entries with missing product_first_sold_date should be dropped


#### Note
For column properties that have been identified as being crucial to our analysis and we do not want to drop entries with empty values for that column, a specified default can be set.
